### ISSUE TITLE
Improve Startup UX & perceived performance

### ISSUE THREAD

**Issue by bryphe** [onivim/oni#355]
The "startup" experience of Oni is very choppy and disjointed. The experience today is:
- ...some delta between running 'oni' and having a browser window show up
- A blank, white screen is shown
- A black screen with a green cursor in the top left is shown
- The editor finally appears

I'd like to address it by the following:
- Show a splash screen as quickly as possible
- Show a black screen with a spinning 'oni' logo in the center of the screen immediately
- Once Oni is loaded, have the editor fade in and the 'oni' logo disappear

Once we have 'spectron' tests, we can start having a benchmark for the load time for Oni across different platforms.

**Comment by keforbes** [comment]
I was thinking earlier about this but with a different approach... would it be possible to launch electron, connect to Neovim, wait maybe 1-2 seconds and *then* load all the fancy features Oni provides?  If we can display a window with Neovim as fast as possible I think it would be fine for a small delay before loading all the quickinfo, autocomplete, fuzzy finder, command palette, etc.

I guess in my mind, if I see the text of a file immediately and then 2 seconds later the syntax highlighting kicks in, that doesn't bother me.  I don't consider that a slow application.  Yet if I sit there for 2 seconds wondering if the OS registered my double-click or if something is hung and suddenly the app appears but is fully loaded, I'd still think it was slow.  Personally, I think "get the file displayed" should be the top priority and everything else can be secondary without sacrificing the "feel" of a quick startup.  It's all user perception anyway; numbers don't matter if things "feel" quick.

**Comment by extr0py** [comment]
Yes, sounds like we're pretty much on the same page here! One benefit of the architecture Oni has - that of running JavaScript plugins outside of Neovim - is that we can load them in parallel while Oni is loading.  As you mention, you really don't need all of the quickinfo/autocomplete immediately, and that can be deferred and parallelized until we've actually rendered some UI.

Another nice thing is that, as we start externalizing the UI (#201, and then perhaps tab menu), is that we can start rendering those items before the Neovim process even starts. 

> I guess in my mind, if I see the text of a file immediately and then 2 seconds later the syntax highlighting kicks in, that doesn't bother me. I don't consider that a slow application. 

Agree 100%. One of the most important metrics we should start tracking is "time to show something" - basically the time from starting to seeing _something_ from the UI. I believe we actually have a lot of room for improvement here, since there has been almost no optimization in the load pipeline currently. https://github.com/electron/spectron seems like a pretty cool tool that can help us get this measurement - if we had an automated test with spectron that measures the time for the first BrowserWindow to show, that could be a good first step. We could make sure that we aren't regressing performance with changes and also that performance improvements actually have a measurable impact. I had this #298 open for a simple validation test (to make sure Oni launches on all platforms), but a good next step would be the performance validation

**Comment by bryphe** [comment]
This should be fully addressed now by #1009 - the work left is to optimize specific paths on startup, like #372 .
